### PR TITLE
Client cert download supports both .pem and .p12 formats

### PR DESCRIPTION
**Pull request by jacobythwaites**
This is required for the Android and iPhone apps and for web browsers.

The `POST /sse_cfg/cert` requires a 3rd parameter `format` which can be either `pem` or `p12`.

### CODE REVIEW

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 1]
```diff
@@ -94,7 +98,7 @@ handle_request(_Arg, 'DELETE',
     "/user/" ++ Delete, #?USER{} = User) ->
   delete_user(Delete, User);
 
-%% Returns the client certificate for the corresponding user, if
+%% Returns the client certificate in for the corresponding user, if
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]: ?

File: apps/sse_cfg/priv/docroot/account/account.html [hunk 2]
```diff
@@ -47,23 +47,15 @@
           <button name="register" title="Use credentials to register a new user">
             Register
           </button>
-          <button name="cert" title="Use credentials to download a client certificate">
+          <button name="cert" title="Use credentials to download client cert in PEM format">
```
Comment (lucafavatellawork) [inline apps/sse_cfg/priv/docroot/account/account.html]: Is this string up-to-date? I had the impression that the format was depending on the format option, while the string mentions "PEM".

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 3]
```diff
@@ -264,36 +270,48 @@ assert_not_self(_Caller, _DeleteUser) ->
...
     sse_cfg:get_user(
       Name, Password),
 
-  {ok, ComboPemFile} =
-    sse_yaws_tls:gen_client_certs(
-      ClaimedUser#?USER.name),
+  {CertFormat, MimeType} =
+    case Format of
+      ?pem -> {pem, "application/x-pem-file"};
+      ?p12 -> {p12, "application/pkcs12"}
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]: At https://www.iana.org/assignments/media-types/media-types.xhtml I see https://www.iana.org/assignments/media-types/application/pkcs12 - ok.

File: apps/sse_cfg/src/sse_cfg_yapp.erl [hunk 4]
```diff
@@ -264,36 +270,48 @@ assert_not_self(_Caller, _DeleteUser) ->
...
   {ok, ClaimedUser} =
     sse_cfg:get_user(
       Name, Password),
 
-  {ok, ComboPemFile} =
-    sse_yaws_tls:gen_client_certs(
-      ClaimedUser#?USER.name),
+  {CertFormat, MimeType} =
+    case Format of
+      ?pem -> {pem, "application/x-pem-file"};
```
Comment (lucafavatellawork) [inline apps/sse_cfg/src/sse_cfg_yapp.erl]: At https://www.iana.org/assignments/media-types/media-types.xhtml the most similar I found is https://www.iana.org/assignments/media-types/application/pem-certificate-chain

Where does `application/x-pem-file` come from?

File: apps/sse_cfg/priv/docroot/account/account.html (hunk 2 above)
Comment (jacobythwaites) [reply apps/sse_cfg/priv/docroot/account/account.html]: Fixed

**Comment by jacobythwaites** [comment]
> * sse_core repo, file Makefile, targets node_cert and clean_node_cert.

Agree, removed.

File: apps/sse_yaws/src/sse_yaws_tls.erl [hunk 5]
```diff
@@ -232,47 +233,84 @@ gen_certbot_certs() ->
...
+
+    PathName =
+      sse_util:format(
+        "{{work_dir}}/{{user_name}}.combo.{{cert_format}}", #{
+          work_dir => WorkDir,
+          user_name => UserName,
+          cert_format => CertFormat}),
+
+    ok =
+      case filelib:is_regular(PathName) of
```
Comment (lucafavatellawork) [inline apps/sse_yaws/src/sse_yaws_tls.erl]: What is meant to happen if a cert had been generated but is now expired, and user now requests a new one? Is the expired one always returned? Does that mean that in that case the user is meant to use cookie forever?

File: apps/sse_yaws/priv/openssl/gen_client_certs.sh [hunk 6]
```diff
@@ -70,5 +71,12 @@ openssl req -new -key ${SSE_CLIENT_KEY} -out ${SSE_CLIENT_REQ} \
...
-cat ${SSE_CLIENT_KEY} ${SSE_CLIENT_CERT} > ${SSE_CLIENT_COMBO}
+# Concatenate the combined client key and cert (.pem file)
+cat ${SSE_CLIENT_KEY} ${SSE_CLIENT_CERT} > ${SSE_CLIENT_COMBO_PEM}
+
+# Generate the matching .p12 version of the .pem file with the
+# fixed password of the user name e.g. 'owner@localhost'.
+openssl pkcs12 -export \
+  -in ${SSE_CLIENT_COMBO_PEM} \
+  -out ${SSE_CLIENT_COMBO_P12} \
+  -passout pass:${USER_NAME}
```
Comment (lucafavatellawork) [inline apps/sse_yaws/priv/openssl/gen_client_certs.sh]: How does the user - downloading this cert - know what the password is?

File: apps/sse_yaws/src/sse_yaws_tls.erl (hunk 5 above)
Comment (jacobythwaites) [reply apps/sse_yaws/src/sse_yaws_tls.erl]: We have no lifecycle implemented for certs yet, except that when a new one is generated (using `sse_yaws_tls:gen_client_certs`) the existing one for that user is revoked in the openssl `.txt` database. I have raised a ticket on the backlog for this: 

https://jira.onfido.co.uk/secure/RapidBoard.jspa?rapidView=155&view=detail&selectedIssue=PID-56

File: apps/sse_yaws/priv/openssl/gen_client_certs.sh (hunk 6 above)
Comment (jacobythwaites) [reply apps/sse_yaws/priv/openssl/gen_client_certs.sh]: The password is a constant value, being unnecessary but unavoidable. 

If the password can be programmatically provided, for example by the user's app, then the end user will of course not have to take any action.

**Pull request merged**